packages/functionals/botpress-qna/src/providers/microsoft_qna_maker.js

```diff
@@ -0,0 +1,171 @@
+import _ from 'lodash'
+import axios from 'axios'
+import ms from 'ms'
+import Promise from 'bluebird'
+
+const KNOWLEDGEBASE_NAME = 'botpress'
+
+// Handles QnA Maker API downcasing all key-values in metadata
+const markUpperCase = str => str.replace(/([A-Z])/g, 'sssooo$1sss')
```

Comment:
I understand the intent here but it's ugly indeed :) Still something along that lines should be done, but I would use more exotic chars.
The 2nd function can be simplified using the function as replacer:

```
const makeLowerCase = str => str.replace(/([A-Z])/g, '##$1##').toLowerCase()
const restoreCase = str => str.replace(/##([a-z])##/g, (_, c) => c.toUpperCase())
```

Reply:
That's the problem - they only allow dashes and underscores. But even for those they don't allow things like `---` or `___` (sequences of special chars).

Reply:
Wow this is impressive. OK let's use some letter combination, just simplify the 2nd method